File: src/python/hiped/penalization.py
import numpy as np
import matplotlib.pyplot as plt
   
class Penalization:
    """
    A class used to represent a penalization function.
    
    Feel free to add your own custom penalization functions to the library in 
    the __init__ method.

    Attributes
    ----------
    
    CoeffPenal : float
        Penalization coefficient
        
    Type : str
        Label of the penalization
        
    Reference : str
        Litterature related to the penalization
        
    Expression : function
        Expression of the penalization function
    
    Derivative : function
        Expression of the derivative of the penalization function
        
    Methods
    -------
    
    eval(x)
        Compute the penalization evaluated on x
        
    evald(x)
         Compute the penalization derivative evaluated on x 
        
    plot(derivative = False)
        Plot the penalization or its derivative on [0,1] interval.
            
    License
    -------
    
    Copyright (C) 2024 Théodore CHERRIÈRE (theodore.cherriere@ricam.oeaw.ac.at)
    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    any later version.
    
    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
    
    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
    """
    
    def __init__(self, penalType = "simp", coeffPenal = 1, reverse = False, reference = "",
                 derivative1 = 1, derivative2 = 1):
        
        self.CoeffPenal = coeffPenal
        self.Type = penalType
        if reverse: 
            self.Type = "1 - " + self.Type + "(1-x)"
            
        
        if penalType.lower() in ["simp", "power", "powerlaw", "power law", "pow", "power_law"]:
            
            self.Reference = "Mlejnek (1992), Bendsøe & Sigmund (1999), Sigmund (2003)"
            self.Expression= lambda x: np.real(x**self.CoeffPenal)
            self.Derivative= lambda x: np.real(self.CoeffPenal * x**(self.CoeffPenal-1))
            if reverse:
                self.Expression= lambda x: 1 - np.real((1-x)**self.CoeffPenal)
                self.Derivative= lambda x: np.real(self.CoeffPenal * (1-x)**(self.CoeffPenal-1))
        
        if penalType.lower() in ["ramp", "r", "rat", "rational"]:
            
            self.Reference = "Stolpe and Svanberg (2001), Hansen (2005)"
            self.Expression= lambda x: np.real(x / (self.CoeffPenal*(1-x) + 1))
            self.Derivative= lambda x: np.real((self.CoeffPenal + 1)/(self.CoeffPenal*(1-x) + 1)**2)
            if reverse:
                self.Expression= lambda x: 1-np.real( (1-x) / (self.CoeffPenal*x + 1))
                self.Derivative= lambda x: - np.real((x*(x-1)) / (self.CoeffPenal*x + 1 )**2)
        
        if penalType.lower() in ["lukas", "lukáš", "atan", "arctan", "heaviside"]:
            
            self.Reference = "Lukáš (2006) - An Integration of Optimal Topology and Shape Design for Magnetostatics"
            self.Expression= lambda x: np.real((1 + np.arctan(self.CoeffPenal*(2*x-1))/np.arctan(self.CoeffPenal))/2)
            self.Derivative= lambda x: np.real(self.CoeffPenal/(np.arctan(self.CoeffPenal) * (self.CoeffPenal**2 * (2*x - 1)**2 + 1)))
            # no change for this penalization type when reversed

        if penalType.lower() in ["zhou", "rational2","rat2", "zhu"]:    
            self.Expression= lambda x: np.real(x/self.CoeffPenal + (self.CoeffPenal-1)/self.CoeffPenal*x**3)
            self.Derivative = lambda x: np.real(1/self.CoeffPenal + (3*x**2*(self.CoeffPenal-1))/self.CoeffPenal)
            self.reference = "don't remember";
            if reverse:
                self.Expression= lambda x: 1 - np.real((1-x)/self.CoeffPenal + (self.CoeffPenal-1)/self.CoeffPenal*(1-x)**3)
                self.Derivative= lambda x: np.real(1/self.CoeffPenal + 3*(self.CoeffPenal - 1)*(x - 1)**2 / self.CoeffPenal)
                
        # To continue with other penalizations
    
    def eval(self, x):
        return self.Expression(x)
    
    def evald(self, x):
        return self.Derivative(x)
    
    
    def plot(self, derivative = False):
        x = np.arange(0,1.01,0.01)
        if not derivative:
            plt.plot(x,self.eval(x), label = f" p = {self.CoeffPenal}")
            plt.ylabel("$P(\omega)$")
        else:
            plt.plot(x,self.deval(x), label = f" p = {self.CoeffPenal}")
            plt.ylabel("$dP/d\omega$")
        plt.grid()
        plt.legend()
        plt.xlabel("$\omega$")
        plt.axis("equal")
        

File: src/matlab/class/Penalization.m
classdef Penalization
     %Penalization Class
    %
    % To create a Penalization object, type
    % obj = Penalization(type,p)
    %
    % - 'type' is a string identifier related to a type of penalization
    % such as "simp", "ramp", etc.
    % - p is the penalization coefficient
    %
    % Penalization are applied to shape functions. They can be useful in
    % the context of topology optimization.
    %
    %Copyright (C) 2024 Théodore CHERRIERE (theodore.cherriere@ricam.oeaw.ac.at.fr)
    %This program is free software: you can redistribute it and/or modify
    %it under the terms of the GNU General Public License as published by
    %the Free Software Foundation, either version 3 of the License, or
    %any later version.
    %
    %This program is distributed in the hope that it will be useful,
    %but WITHOUT ANY WARRANTY; without even the implied warranty of
    %MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    %GNU General Public License for more details.
    %
    %You should have received a copy of the GNU General Public License
    %along with this program.  If not, see <https://www.gnu.org/licenses/>.
    properties
        Expression % function handle
        Derivative % function handle
        Type       % str
        Reference  % str
        Coeff      % double
    end

    methods
        function obj = Penalization(varargin)
            % obj = Penalization(type, p)
            % - type is a string identifier refering to a type of
            % penalization (available : "simp", "ramp", "zhu", "lukas", 
            % and some of their concave conterpart)
            % - p is the penalization coefficient

            type = varargin{1};
            if nargin<=1 || isempty(varargin{2})
                obj.Coeff=1;
            else
                obj.Coeff=varargin{2};
            end
            if lower(type)=="simp"
                obj.Expression=@(x,p) real(x.^p);
                obj.Derivative=@(x,p) real(p*x.^(p-1));
                obj.Type=lower(type);
                obj.Reference = "Mlejnek (1992), Bendsøe and Sigmund (1999), Sigmund (2003)";

            elseif lower(type)=="simp_concave" || lower(type)=="simp_inv"
                obj.Expression= @(x,p) 1-(real((1-x).^p));
                obj.Derivative=@(x,p) real(p.*(1 - x).^(p - 1));
                obj.Type=lower(type);
                obj.Reference = "1-simp(1-x,p)";

            elseif lower(type)=="ramp"
                obj.Expression= @(x,p) real(x./(p - p.*x + 1));
                obj.Derivative=@(x,p)real((p + 1)./(p - p.*x + 1).^2);
                obj.Type=lower(type);
                obj.Reference = "Stolpe and Svanberg (2001), Hansen (2005)";

            elseif lower(type)=="ramp_concave" || lower(type)=="ramp_inv"
                obj.Expression= @(x,p) 1-real((1-x)./(p - p.*(1-x) + 1));
                obj.Derivative=@(x,p) real(1./(p.*x + 1)) - real((p.*(x - 1))./(p.*x + 1).^2);
                obj.Type=lower(type);
                obj.Reference = "1-ramp(1-x,p)";

            elseif lower(type)=="lukas"
                obj.Expression= @(x,p) real((1+atan(p*(2*x-1))./atan(p))/2);
                obj.Derivative = @(x,p) real(p./(atan(p).*(p.^2.*(2*x - 1).^2 + 1)));
                obj.Type=lower(type);
                obj.Reference = "Lukáš (2006) - An Integration of Optimal Topology and Shape Design for Magnetostatics";
            else
                error("unknown penalization type")
            end
        end

        function out = eval(obj,x)
            % out = eval(obj,x)
            % evaluate the penalization (x should be in [0,1])
            out = obj.Expression(x,obj.Coeff);
        end

        function out = evald(obj,x)
            % evaluate the penalization derivative (x should be in [0,1])
            out = obj.Derivative(x,obj.Coeff);
        end

        %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
        %% Display
        function plot(obj)
            % plot(obj)
            % plot the penalization function in [0,1]
            x=0:0.01:1;
            plot(x,obj.eval(x));
            grid on
            xlabel("\omega")
            ylabel("P(\omega)")
        end
    end
end

